main/lsp/LSPTypecheckerCoordinator.cc

```diff
@@ -8,33 +8,65 @@ using namespace std;
 LSPTypecheckerCoordinator::LSPTypecheckerCoordinator(const shared_ptr<const LSPConfiguration> &config)
     : shouldTerminate(false), typechecker(config), config(config), hasDedicatedThread(false) {}
 
-void LSPTypecheckerCoordinator::asyncRunInternal(function<void()> &&lambda) {
+void LSPTypecheckerCoordinator::asyncRunInternal(function<void()> &&lambda, bool canPreemptSlowPath) {
     if (hasDedicatedThread) {
+        // We can only preempt if this lambda will run next. We have no notion of a queue of preemption lambdas.
+        if (canPreemptSlowPath && lambdas.enqueuedEstimate() - lambdas.doneEstimate() == 0) {
```

Comment:
Note: I think I need to introduce a new method to the queue that doesn't access these atomics in a relaxed manner. I need accuracy here.